auditbeat/helper/hasher/hasher.go

```diff
@@ -0,0 +1,264 @@
+// under the License.
+
+package hasher
+
+import (
+	"crypto/md5"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/hex"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/OneOfOne/xxhash"
+	"github.com/dustin/go-humanize"
+	"github.com/joeshaw/multierror"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/sha3"
+	"golang.org/x/time/rate"
+
+	"github.com/elastic/beats/libbeat/common/file"
+)
+
+// HashType identifies a cryptographic algorithm.
+type HashType string
+
+// Unpack unpacks a string to a HashType for config parsing.
+func (t *HashType) Unpack(v string) error {
+	*t = HashType(strings.ToLower(v))
+	return nil
+}
+
+// IsValid checks if the hash type is valid.
+func (t *HashType) IsValid() bool {
+	_, valid := validHashes[*t]
+	return valid
+}
+
+var validHashes = map[HashType](func() hash.Hash){
+	BLAKE2B_256: func() hash.Hash {
+		h, _ := blake2b.New256(nil)
+		return h
+	},
+	BLAKE2B_384: func() hash.Hash {
+		h, _ := blake2b.New384(nil)
+		return h
+	},
+	BLAKE2B_512: func() hash.Hash {
+		h, _ := blake2b.New512(nil)
+		return h
+	},
+	MD5:        md5.New,
+	SHA1:       sha1.New,
+	SHA224:     sha256.New224,
+	SHA256:     sha256.New,
+	SHA384:     sha512.New384,
+	SHA512:     sha512.New,
+	SHA512_224: sha512.New512_224,
+	SHA512_256: sha512.New512_256,
+	SHA3_224:   sha3.New224,
+	SHA3_256:   sha3.New256,
+	SHA3_384:   sha3.New384,
+	SHA3_512:   sha3.New512,
+	XXH64: func() hash.Hash {
+		return xxhash.New64()
+	},
+}
+
+// Enum of hash types.
+const (
+	BLAKE2B_256 HashType = "blake2b_256"
+	BLAKE2B_384 HashType = "blake2b_384"
+	BLAKE2B_512 HashType = "blake2b_512"
+	MD5         HashType = "md5"
+	SHA1        HashType = "sha1"
+	SHA224      HashType = "sha224"
+	SHA256      HashType = "sha256"
+	SHA384      HashType = "sha384"
+	SHA3_224    HashType = "sha3_224"
+	SHA3_256    HashType = "sha3_256"
+	SHA3_384    HashType = "sha3_384"
+	SHA3_512    HashType = "sha3_512"
+	SHA512      HashType = "sha512"
+	SHA512_224  HashType = "sha512_224"
+	SHA512_256  HashType = "sha512_256"
+	XXH64       HashType = "xxh64"
+)
+
+// Digest is a output of a hash function.
+type Digest []byte
+
+// String returns the digest value in lower-case hexadecimal form.
+func (d Digest) String() string {
+	return hex.EncodeToString(d)
+}
+
+// MarshalText encodes the digest to a hexadecimal representation of itself.
+func (d Digest) MarshalText() ([]byte, error) { return []byte(d.String()), nil }
+
+// FileTooLargeError is the error that occurs when a file that
+// exceeds the max file size is attempting to be hashed.
+type FileTooLargeError struct {
+	fileSize int64
+}
+
+// Error returns the error message for FileTooLargeError.
+func (e FileTooLargeError) Error() string {
+	return fmt.Sprintf("hasher: file size %d exceeds max file size", e.fileSize)
+}
+
+// Config contains the configuration of a FileHasher.
+type Config struct {
+	HashTypes           []HashType `config:"hash_types,replace"`
```

Comment:
q: What's does the `,replace` option do? I haven't see that used before.

Reply:
`replace` allows `process.hash.hash_types` to be empty and disable hashing altogether that way (e.g. if there is a performance problem, or simply to restore current behavior). Only when the config option is not present at all does the default (`sha1`) take effect.

Without `replace`, the default value will be used even when the option is present but empty, so there would be no way of disabling it.

I would prefer something called `allow_empty` or similar, but `replace` is what I found works as well.

@urso maybe you can take a quick look at whether this is the right use.